chore(governance): add pre-configured dev environment with GitPod.io to ease contributions

leandrodamascena opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** #1361 

## Summary
Added GitPod.io support for external contributions to make creating a development environment easier

### Changes

> Please provide a summary of what's being changed

Added integration with Gitpod.io and through the badge/link in README.md it will be very simple for any contributor to configure a new environment

![image](https://user-images.githubusercontent.com/4295173/182268971-c39aa9f1-8bdb-4802-a447-c5780080b47e.png)


### User experience

> Please share what the user experience looks like before and after this change

Before: any contributor had to clone the project, install dependencies, and set up an environment in order to start development. This generates an overhead for new people who want to contribute to the project, as even a small fix required setting up a full new environment.

After: any contributor just needs to create an environment on Gitpod using the fork URL and everything will be set up.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.


-----
[View rendered CONTRIBUTING.md](https://github.com/leandrodamascena/aws-lambda-powertools-python/blob/chore/gitpod-dev/CONTRIBUTING.md)
[View rendered README.md](https://github.com/leandrodamascena/aws-lambda-powertools-python/blob/chore/gitpod-dev/README.md)

leandrodamascena committed:
chore(dev): adding gitpod support

leandrodamascena committed:
chore(docs): adding gitpod badge

leandrodamascena committed:
chore(docs): adding gitpod section to CONTRIBUTING.md

leandrodamascena commented on CONTRIBUTING.md:
```diff
@@ -36,6 +36,8 @@ To send us a pull request, please follow these steps:
 5. Send us a pull request with a [conventional semantic title](https://github.com/awslabs/aws-lambda-powertools-python/pull/67), and answering any default questions in the pull request interface.
 6. Pay attention to any automated CI failures reported in the pull request, and stay involved in the conversation.
 
+Optionally you can use [Gitpod](https://gitpod.io/) to have a pre-configured dev environment. If you choose to use Gitpod, there is no need to install anything on your machine, as the environment will be pre-configured with all necessary dependencies.  
```
Hello @heitorlessa and @rubenfonseca. I'm really not sure if this information is enough to make it clear how to use Gitpod.io or we should create a section to explain how Gitpod works and how to use it. I need help :pray:

rubenfonseca replied:
Hi Leandro! This is great, but I'm not familiar with gitpod, and after reading this paragraph I don't know what I have to do exactly. Do I need to click somewhere? Do I have to go to gitpod.io instead? Some extra instructions would help a lot :)

leandrodamascena replied:
Hello Rubens! Currently we need to fork the repository, install all dependencies locally and start coding, basically using Gitpod we just need to fork the repository and click on that [badge](https://github.com/awslabs/aws-lambda-powertools-python/pull/1403/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R6) and everything is ready to start coding.
To contribute to the project this document guides us through 7 steps and using Gitpod basically changes step 2 because we no longer need to install dependencies locally.

What do you think about this:

`Optionally you can use [Gitpod](https://gitpod.io/) to have a pre-configured dev environment. If you choose to use Gitpod, there is no need to install anything on your machine, as the environment will be pre-configured with all necessary dependencies. To gain access to a pre-configured environment using Gitpod, simply fork the repository and click the Gitpod badge in the README.md file.`

heitorlessa replied:
I'd suggest updating the entire section and use something along these lines. This makes our recommendation to use our pre-configured env explicit, give them recommendations on how to create against their fork (not ours, or else their push will fail), and keeping the remaining steps applicable for both.

The only thing I kinda miss is having a link for people unfamiliar with GitHub trying to create their first PR - maybe installing GitHub CLI and adding [this link](https://cli.github.com/manual/examples.html) can raise the bar? I'm fine without it for now too.

---

Firstly, [fork the repository](https://github.com/awslabs/aws-lambda-powertools-python/fork).

To setup your development environment, we recommend using our pre-configured Cloud environment: https://gitpod.io/#https://github.com/YOUR_USERNAME/aws-lambda-powertools-python. Replace `YOUR_USERNAME` with your GitHub username so the Cloud environment can target your fork accordingly.

Alternatively, you can use `make dev` within your local virtual environment.

To send us a pull request, please follow these steps:

1. Create a new branch to focus on the specific change you are contributing e.g. `improv/logger-debug-sampling`
2. Run all tests, and code baseline checks: `make pr`
    - Git hooks will run linting and formatting while `make pr` run deep checks that also run in the CI process
3. Commit to your fork using clear commit messages.
4. Send us a pull request with a [conventional semantic title](https://github.com/awslabs/aws-lambda-powertools-python/pull/67), and answering any default questions in the pull request interface.
5. Pay attention to any automated CI failures reported in the pull request, and stay involved in the conversation.

rubenfonseca commented on .gitpod.yml:
```diff
@@ -0,0 +1,11 @@
+    - littlefoxteam.vscode-python-test-adapter
+    - ms-azuretools.vscode-docker
+    - davidanson.vscode-markdownlint
+    - bungcip.better-toml
```
Can you configure your editor to add a new line at the end of a file? This avoids unnecessary diffs in the future

https://thoughtbot.com/blog/no-newline-at-end-of-file#following-the-rules-in-your-editor

leandrodamascena replied:
Hi Rubens, added!

rubenfonseca commented on .gitpod.Dockerfile:
```diff
@@ -0,0 +1,10 @@
+
+# Installing pre-commit as system package and not user package. Git needs this to execute pre-commit hooks.
+RUN export PIP_USER=no
+RUN python3 -m pip install pre-commit
```
Can you configure your editor to add a new line at the end of a file? This avoids unnecessary diffs in the future

[thoughtbot.com/blog/no-newline-at-end-of-file#following-the-rules-in-your-editor](https://thoughtbot.com/blog/no-newline-at-end-of-file#following-the-rules-in-your-editor)

leandrodamascena replied:
Hi Rubens, added!

rubenfonseca reviewed (changes requested):
Awesome first PR @leandrodamascena, so happy for this!! Looking forward to try gitpod too and to ease the barrier for contribution. Thanks so much!

leandrodamascena committed:
chore(dev): adding new lines at the end of the file

heitorlessa commented on Makefile:
```diff
@@ -9,6 +9,11 @@ dev:
 	pre-commit install
 
+dev-gitpod:
+	pip install --upgrade pip poetry
```
python3 -m pip?

heitorlessa replied:
One thing to verify is whether additional dependencies will work as expected if anyone does: `poetry add a_library --dev`

leandrodamascena replied:
Hi @heitorlessa.
1 - I think it's better to use pip directly instead of the python3 pip module. The only reason I used python3 -m pip to install pre-commit is because git looks for the system path to python (/home/gitpod...) when running pre-commit hook. To install poetry and extra dependencies, I think it more suitable to use pip directly because it will install dependencies in the user's workspace (/workspace...).

2 - It makes a lot of sense to install a new dependency and test if it's working as expected, good catch. Here is the screenshots installing a new dependency and using this. Please disregard the purpose of this test, it's just to demonstrate that it works.

![image](https://user-images.githubusercontent.com/4295173/182372277-49b5c882-5157-432e-b69a-dc158686982e.png)
![image](https://user-images.githubusercontent.com/4295173/182372468-29042898-4f36-42d5-9ec7-2b7d9dad0a21.png)
![image](https://user-images.githubusercontent.com/4295173/182372537-0589af30-04e5-4365-a4a2-0c2a2bcfbea6.png)
![image](https://user-images.githubusercontent.com/4295173/182372568-7bde1098-7277-4c1e-8c0f-ff716b886fcb.png)
![image](https://user-images.githubusercontent.com/4295173/182372619-5de9b38d-3391-4a08-b24a-451c811d357b.png)
![image](https://user-images.githubusercontent.com/4295173/182372666-9999e37a-4df8-4ee8-991d-1eefdc26e942.png)

heitorlessa commented on .gitpod.yml:
```diff
@@ -0,0 +1,11 @@
+tasks:
+  - init: make dev-gitpod
+vscode:
+  extensions:
```
Don't recall if they were ported to the open vscode extension, but equally helpful for everyone starting:

- oderwat.indent-rainbow
- yzhang.markdown-all-in-one  # autoformat, better visualization, snippets, and markdown export to multiple fmts
- bierner.markdown-mermaid # previews mermaid diagrams when previewing markdown
- matangover.mypy # highlight mypy issues
- njpwerner.autodocstring # auto-generate docsstrings in numpy format that we use
- netcorext.uuid-generator # for those helping create code snippets for docs
- streetsidesoftware.code-spell-checker # spell checker that works with camel case too
- bungcip.better-toml # in case GitPod doesn't have support for TOML `pyproject.toml`

heitorlessa commented on .gitpod.Dockerfile:
```diff
@@ -0,0 +1,10 @@
+# See here all gitpod images available: https://hub.docker.com/r/gitpod/workspace-python-3.10/tags
+# Current python version: 3.10.5
+FROM gitpod/workspace-python-3.10@sha256:8a7ad4f0bbaa281a36cf2a87b772354638a14d7383f0a755b9ea32596ee99632
```
Let's use 3.9. It's the latest Lambda runtime supports, we can upgrade to 3.10 when they do. Also deeply appreciate the security mindset in using an immutable image

heitorlessa commented on README.md:
```diff
@@ -3,6 +3,7 @@
 ![Build](https://github.com/awslabs/aws-lambda-powertools/workflows/Powertools%20Python/badge.svg?branch=master)
 [![codecov.io](https://codecov.io/github/awslabs/aws-lambda-powertools-python/branch/develop/graphs/badge.svg)](https://app.codecov.io/gh/awslabs/aws-lambda-powertools-python)
 ![PythonSupport](https://img.shields.io/static/v1?label=python&message=3.6%20|%203.7|%203.8|%203.9&color=blue?style=flat-square&logo=python) ![PyPI version](https://badge.fury.io/py/aws-lambda-powertools.svg) ![PyPi monthly downloads](https://img.shields.io/pypi/dm/aws-lambda-powertools)
+[![Gitpod Ready-to-Code](https://img.shields.io/badge/Gitpod-Ready--to--Code-blue?logo=gitpod)](https://gitpod.io/from-referrer/)
```
best to keep in the CONTRIBUTING only, and also in the PR via the GitHub App we can install after the PR merge.

heitorlessa replied:
```suggestion
```

heitorlessa reviewed (changes requested):
some last adjustments before settling on wording within contributing docs.

leandrodamascena committed:
chore(dev): adding new lines at the end of the file

leandrodamascena committed:
chore(dev): changed python version from 3.10.5 to 3.9.13

leandrodamascena committed:
chore(dev): added new vscode extensions to make the life easier

leandrodamascena committed:
chore(dev): fixed typo in python version

leandrodamascena committed:
chore(docs): added gitpod badge in CONTRIBUTING and removed from README

rubenfonseca reviewed (approved):
Thank you! I believe Heitor will have some comments about wording, but otherwise this is nice :) Looking forward for more!

leandrodamascena committed:
chore(docs): adding more details to CONTRIBUTING guide

heitorlessa commented on CONTRIBUTING.md:
```diff
@@ -15,31 +17,33 @@ reported the issue. Please try to include as much information as you can.
 1. You are working against the latest source on the **develop** branch.
 2. You check existing open, and recently merged pull requests to make sure someone else hasn't addressed the problem already.
-3. You open a [RFC issue](https://github.com/awslabs/aws-lambda-powertools-python/issues/new?assignees=&labels=RFC%2C+triage&template=rfc.md&title=RFC%3A+) to discuss any significant work - we would hate for your time to be wasted. -->
+3. You open a [RFC issue](https://github.com/awslabs/aws-lambda-powertools-python/issues/new?assignees=&labels=RFC%2Ctriage&template=rfc.yml&title=RFC%3A+TITLE) to discuss any significant work - we would hate for your time to be wasted.
```
```suggestion
3. You open an [issue](https://github.com/awslabs/aws-lambda-powertools-python/issues/new/choose) before you begin any implementation. We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.
```

heitorlessa replied:
This will match our current disclaimers, as we've updated during the pause.

heitorlessa committed:
chore: sync disclaimer wording for new contributions

leandrodamascena commented:
> Thank you! I believe Heitor will have some comments about wording, but otherwise this is nice :) Looking forward for more!

Thank you @rubenfonseca. It will be really cool to be able to contribute more to this project that I started as a user. :1st_place_medal:

heitorlessa commented:
Thank you so much @leandrodamascena for taking it this far (and addressing all feedbacks). Merging!

leandrodamascena commented:
> Thank you so much @leandrodamascena for taking it this far (and addressing all feedbacks). Merging!

Thanks also @heitorlessa  for insights and helping me make this PR even better. :airplane:
